### PR TITLE
test(negative): implement node network disconnection

### PR DESCRIPTION
**Pull request by yangchiu**
test: implement node network disconnection

For https://github.com/longhorn/longhorn/issues/6704

Signed-off-by: Yang Chiu <yang.chiu@suse.com>

### CODE REVIEW

**Comment by yangchiu** [comment]
blocked by https://github.com/longhorn/longhorn-tests/pull/1542

**Review by c3y1huang — CHANGES_REQUESTED** [review]
There are overlappings with https://github.com/longhorn/longhorn-tests/pull/1542. I will review this again after that one is merged.

**Comment by yangchiu** [comment]
@longhorn/qa Please help review this. Many thanks.

File: e2e/libs/volume/volume.py [hunk 1]
```diff
@@ -49,8 +49,8 @@ def wait_for_volume_state(self, volume_name, desired_state):
         return self.volume.wait_for_volume_state(volume_name, desired_state)
 
     def wait_for_volume_attached(self, volume_name):
-        self.volume.wait_for_volume_state(volume_name, "attached")
         self.volume.wait_for_volume_robustness_not(volume_name, "unknown")
+        self.volume.wait_for_volume_state(volume_name, "attached")
```
Comment (khushboo-rancher) [inline e2e/libs/volume/volume.py:53]: I assume, making sequence change here makes more sense :D

File: e2e/tests/network_disconnect.robot [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    Given Create statefulset 0 using RWO volume
+    FOR    ${i}    IN RANGE    ${LOOP_COUNT}
+        And Keep writing data to pod of statefulset 0
+        When Disconnect volume node network of statefulset 0 for 10 seconds
+        And Wait for volume of statefulset 0 healthy
+        And Wait for statefulset 0 pods stable
+        Then Check statefulset 0 works
+    END
+
+Disconnect Volume Node Network For More Than Pod Eviction Timeout While Workload Heavy Writing
```
Comment (khushboo-rancher) [inline e2e/tests/network_disconnect.robot:29]: Just curious, can we rename test name little bit shorter like `Disconnect Volume Node Network For Long Time While Workload Heavy Writing` and we can add a description to test to explain the pod eviction timeout

**Review by khushboo-rancher — COMMENTED** [review]
LGTM.

@yangchiu Are we planning to add test cases for node network disconnection when replica rebuilding is in progress?

File: e2e/libs/volume/volume.py (hunk 1 above)
Comment (yangchiu) [reply e2e/libs/volume/volume.py:53]: The original sequence order could fail test cases in some conditions. Like the volume is automatically attached, detached and re-attached in a flash.

**Comment by yangchiu** [comment]
> LGTM.
> 
> @yangchiu Are we planning to add test cases for node network disconnection when replica rebuilding is in progress?

I'll add more test cases after these keyword implementation PRs are merged.

File: e2e/tests/network_disconnect.robot (hunk 2 above)
Comment (yangchiu) [reply e2e/tests/network_disconnect.robot:29]: Yes, it's feasible. But I prefer a test case name explains itself. WDYT @longhorn/qa ?

Comment (chriscchien) [reply e2e/tests/network_disconnect.robot:29]: I'm fine with either.

File: e2e/keywords/network.resource [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+Library             ../libs/keywords/network_keywords.py
+Library             ../libs/keywords/workload_keywords.py
+Library             ../libs/keywords/volume_keywords.py
+
+*** Variables ***
+
+
+*** Keywords ***
+Disconnect volume node network of statefulset ${idx} for ${disconnection_time_in_sec} seconds
```
Comment (c3y1huang) [inline e2e/keywords/network.resource:12]:
```suggestion
Disconnect volume node network of statefulset ${statefulset_id} for ${duration} seconds
```

File: e2e/libs/volume/crd.py [hunk 4]
```diff
@@ -214,7 +214,6 @@ def wait_for_volume_state(self, volume_name, desired_state):
             except Exception as e:
                 logging(f"Getting volume {self.get(volume_name)} status error: {e}")
             time.sleep(self.retry_interval)
-        assert self.get(volume_name)["status"]["state"] == desired_state
```
Comment (c3y1huang) [inline e2e/libs/volume/crd.py:217]: Is there a specific reason for the removal of the final assertion? Same below.

**Review by c3y1huang — CHANGES_REQUESTED** [review]

File: e2e/keywords/network.resource (hunk 3 above)
Comment (yangchiu) [reply e2e/keywords/network.resource]: Refined. Thanks!

File: e2e/libs/volume/crd.py (hunk 4 above)
Comment (yangchiu) [reply e2e/libs/volume/crd.py:217]: Refined. Thanks!

**Review by khushboo-rancher — APPROVED** [review]

**Pull request merged**